### PR TITLE
MAINT: Remove duplicated logic between array_wrap and array_prepare

### PR DESCRIPTION
**Pull request by eric-wieser**
This behaves exactly as before

Part of the cleanup needed to fix  #10450

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -243,27 +269,13 @@ _find_array_prepare(PyObject *args, PyObject *kwds,
             obj = PyDict_GetItem(kwds, npy_um_str_out);
```
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: You are on purpose not yet checking whether the `out` argument is a tuple, correct?

**Review by mhvk — APPROVED** [review]
Only a question about whether you'd not rather check for a tuple here as well. Fine to do that in a follow-up PR, though.

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: I'm on purpose just leaving it as it was - I want to push the `out` argument parsing way upstream

**Comment by eric-wieser** [comment]
Actually, I should try to keep the comments from the function I removed - so don't merge this yet

**Comment by eric-wieser** [comment]
Updated with comments and slightly more efficient code.

File: numpy/core/src/umath/ufunc_object.c [hunk 2]
```diff
@@ -184,6 +184,45 @@ _find_array_method(PyObject *args, int nin, PyObject *method_name)
...
 
+/*
+ * Returns an incref'ed pointer to the proper __array_prepare__/__array_wrap__
+ * method for a ufunc output argument, given the output argument `obj`, and the
+ * method chosen from the inputs `input_method`.
+ */
+static PyObject *
+_get_output_array_method(PyObject *obj, PyObject *method,
+                         PyObject *input_method) {
+    if (obj == Py_None) {
```
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: There are now three times in which you have the incref and return of `input_method`. How about inverting the logic, and move the next two lines to the very end, as the fall-back opion, i.e.,
```
PyObject *ometh;
if obj != Py_None` {
    if (PyArrayCheckExact(obj)) {
        return Py_RETURN_NONE;
    }
    ometh = ...
    if (ometh) {
        if(PyCallable_Check(ometh)) {
            return ometh;
        }
        else {
            Py_DECREF(ometh);
        }
    }
    else {
        PyErr_Clear();
    }
}
Py_XINCREF(input_method);
return input_method;
```

**Comment by eric-wieser** [comment]
Should be ready

**Comment by mhvk** [comment]
All OK now, merging!

**Pull request merged**